src/granite/comparator.cr

```diff
@@ -0,0 +1,32 @@
+# Adds equals, hash, and comparison methods to Granite::Base objects.
+module Granite::Comparator
+  macro __process_comparator
+
+    def_equals_and_hash {{ *FIELDS.keys }}
+
+    {%
+      pk_type = PRIMARY[:type].id
+      pk_type_node = PRIMARY[:type].resolve
+      pk_field = PRIMARY[:name].id
+    %}
+
+    {% if pk_type_node.has_method?("<=>") %}
+    # Compares this `{{@type}}` to another `{{@type}}` by comparing their primary keys (`{{pk_field}}`).
+    #
+    # Returns 0 if the two primary keys are equal, a negative number if this object's primary key is considered less than other's, or a positive number otherwise.
+    {% else %}
+    # Raises an exception; the `{{pk_field}}` field does not implement the `<=>` operator.
+    {% end %}
```

Comment:
I think the code below is pretty straightforward and doesn't need clarifying comments, so this can be removed.